[wallet] restore coinbase and confirmed/conflicted checks in SubmitMemoryPoolAndRelay()

jnewbery opened the pull request:
These checks don't change mempool acceptance/relay behaviour, but reduce log spam.

jnewbery committed:
[wallet] restore coinbase check in SubmitMemoryPoolAndRelay()

This check doesn't change mempool acceptance/relay behaviour, but reduces log spam.

jnewbery commented:
Reported by @kallewoof 

ping @MarcoFalke and @ariard for review.

ariard commented on src/wallet/wallet.cpp:
```diff
@@ -2160,6 +2160,9 @@ bool CWalletTx::SubmitMemoryPoolAndRelay(std::string& err_string, bool relay)
     if (isAbandoned()) return false;
+    // Don't try to submit coinbase transactions. These would fail anyway but would
+    // cause log spam.
+    if (IsCoinBase()) return false;
```
Sorry for that. You may need also to restore the `GetDepthInMainChain`, if you follow what @kallewoof was saying non-coinbase confirmed txn were also in log spams.

kallewoof replied:
No, I don't believe non-coinbase confirmed txn were in log spams but there were thousands so I can't easily verify.

ariard reviewed (commented):
Once wallet rebroadcast logic get reworked, I think log spams due to massive rebroadcast of confirmed txn should disappear, but yes that's not a thing for now.

promag commented:
Related discussion in #https://github.com/bitcoin/bitcoin/pull/15713/files#r308801356.

ACK 214c4ecb9ab306627a08abb35cf82c73f10ec627.

ariard commented:
ACK 214c4ec

kallewoof commented:
I applied this patch to my node and restarted it.

1. This does not only happen during reorgs, it seems to be happening regularly. I think the node has gotten in its head that it needs to rebroadcast *all* transactions.
2. This is not only coinbase transactions. After applying the patch I see tons of txs. Picking one at random, it turned out to be a regular tx. See https://explorer.bc-2.jp/tx/a2ba0e7b09474565c92ef495a35fd932691f3ee3468992cd6a7e5bce100a6b3a (`2019-08-07T08:53:22Z [default wallet] Submitting wtx a2ba0e7b09474565c92ef495a35fd932691f3ee3468992cd6a7e5bce100a6b3a to mempool for relay`)
3. This did not happen in conjunction with a new block, either. It seems the wallet is doing this as part of its regular wallet tx resend. But these have thousands of confirmations.

ariard commented:
Did you try to run without #15713 but with `IsCoinBase` and `GetDepthInMainChain` checks turn off ? I guess you should see the same logs than right now. Wallet is always rebroadcasting, even with thousands of confirmations, removing checks just make it apparent ?

kallewoof commented:
@ariard No. If I am the only one seeing these messages, however, this may be a problem with signet. No clue why signet would cause this, though, tbh.

MarcoFalke commented:
#15713 also changed the logging, IIRC. So the bug can be fixed by either (or both):

* Fixed logging
* Restored checks

ariard commented:
@kallewoof is not a problem with signet, it's a problem with my PR. I've played a bit with a regtest node on a commit before it. If I generate thousands of coinbase txn, every time we hit `nNextResend` (once in a 30min period), and 5min older than last block we are going to call `RelayWalletTransaction` for every one of them. You can see them in this method if you add a printer before `IsCoinBase` check. Previously printer was only triggered if they have been accepted by mempool. Now, with the new refactored `SubmitMemoryPoolAndRelay` we have removed these checks, because they are already performed in `BroadcastTransaction` or lower, and log print whatever their mempool acceptance/relay are. The resend logic has always been there for unconfirmed txn, it's just now we see it. 

While waiting for this logic to be rationalized, to avoid spam we can move log print to `BroadcastTransaction` after inserting successfully in mempool or re-add both `IsCoinBase` and `GetDepthInMainChain` checks in `SubmitMemoryPoolAndRelay`.

Note: IMO, comment "only rebroadcast unconfirmed txes older than 5 minutes before the last block was found" in `ResendWalletTransactions` is ambiguous and suggests confirmation check has already been done at this step, but in fact it relies on the callee function to do so.

jnewbery committed:
[wallet] Restore confirmed/conflicted tx check in SubmitMemoryPoolAndRelay()

Restores the confirmed/conflicted tx check removed in
8753f5652b4710e66b50ce87788bf6f33619b75a. There should be no external
behaviour change (these txs would not get accepted to the mempool
anyway), but not having the check in the wallet causes log spam.

Also adds a comment to ResentWalletTransactions() that
confirmed/conflicted tx check is done in SubmitMemoryPoolAndRelay().

jnewbery commented:
I've restored the `GetDepthInMainChain()` check in `SubmitMemoryPoolAndRelay()`. That meant passing the `locked_chain` through to `SubmitMemoryPoolAndRelay()` (at least until #15931 is merged).

Also clarified comment in `ResendWalletTransactions()`.

MarcoFalke commented:
ACK c8b53c3beafa289dcdbd8c2ee9f957bdeca79cbc (non-doc changes are mostly a git revert 8753f5652b4710e66b50ce87788bf6f33619b75a)

<details><summary>Show signature and timestamp</summary>

Signature:

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA512

ACK c8b53c3beafa289dcdbd8c2ee9f957bdeca79cbc (non-doc changes are mostly a git revert 8753f5652b4710e66b50ce87788bf6f33619b75a)
-----BEGIN PGP SIGNATURE-----

iQGzBAEBCgAdFiEE+rVPoUahrI9sLGYTzit1aX5ppUgFAlwqrYAACgkQzit1aX5p
pUgQAQv/Z6Y5v5REeQFDBT4A4iyBncehSPa36ikjjvjE1SRHw8gMLoYffjC96s0n
aZ1vpEUrUPhnQhiOWmr68rEFoAkoW13FeMqdc7iN4UBcW71vX2vGVrDmQilqrkid
TcTdogmucTkRCXyrkU6fjSDgO9TAUoF4ANZNuwusWW9oxcQurt8/86PAGI3jAxmy
yEYTmjoWt55EcnzC57rlMQ7mCR+okiEtgV+pC5tXerJvjdCZRM4LqXhYbto1UtOm
i6Qxn6gp4Z1eRjM86KQ6D9YfsP1sxtdS1xVDLXH2JbCD1EB7kE1FzB/IveFyYg3x
56sBd9tenZ8F6UKbhbbnvfDB/AiZwEnkYyBxclB7fVJsH+2N80pGJMcqUwlKBP3y
bHBr+IJhekG70belcOTK02eoWC9OJR0Ga5cNGJwIKmWPScHItnhjC0lzjE4PO0yA
+aJAnuQpeySSfD9TvunAfPxqQ/4EvYk52WOtBxdqHPbNnr0I7Rhtbk0GKaZYWpn3
O4uPQ/hf
=TlR4
-----END PGP SIGNATURE-----
```

[Timestamp of file with hash `d299ae4192e52358020d1c810082e7d19625bf7a9adef077079a3e02624e9dd6  -`](https://opentimestamps.org/info/?004f70656e54696d657374616d7073000050726f6f6600bf89e2e884e892940108d299ae4192e52358020d1c810082e7d19625bf7a9adef077079a3e02624e9dd6f010c76ea1979f6acefa491b1c0bec9b3c9908fff0101e34f57cf839ad9703d70b934b87867a08f1045d4da46df008407e1ab46130a1bf0083dfe30d2ef90c8e232268747470733a2f2f6274632e63616c656e6461722e636174616c6c6178792e636f6dfff0105adb4a1aa10c5a620296f804b8b430ca08f1045d4da46ef008d711d73909b18b230083dfe30d2ef90c8e292868747470733a2f2f66696e6e65792e63616c656e6461722e657465726e69747977616c6c2e636f6dfff0105b8ac417acba5f3cc53c6f17633ca48208f1045d4da46ef00801600bbc7fa11bfa0083dfe30d2ef90c8e2e2d68747470733a2f2f616c6963652e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267f0105ba082009be43279626a3f97bef3b35a08f1045d4da46df008a17a92af0a1e3ed90083dfe30d2ef90c8e2c2b68747470733a2f2f626f622e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267)

</details>

ariard commented:
utACK c8b53c3

kallewoof commented:
Will apply this to my node and see how it goes.

Sjors commented on src/wallet/wallet.cpp:
```diff
@@ -2377,11 +2379,12 @@ void CWallet::ResendWalletTransactions()
-            // last block was found
+            // Attempt to rebroadcast all txes more than 5 minutes older than
+            // the last block. SubmitMemoryPoolAndRelay() will not rebroadcast
+            // any confirmed or conflicting txs.
```
Why not move the coinbase check here then?

jnewbery replied:
I agree that would make sense - the other callers of `SubmitMemoryPoolAndRelay()` do checks before calling into `SubmitMemoryPoolAndRelay()`. However, this PR is just to restore previous behaviour. There are a few other PRs in progress on the node/wallet interface, so let's keep the existing behaviour and look for ways to tidy it up later.
